contracts/helpers/TransferHelper.sol

```diff
@@ -70,97 +88,242 @@ contract TransferHelper is TransferHelperInterface, TokenTransferrer {
+
+        // Derive the conduit address from the deployer, conduit key
+        // and creation code hash.
+        address conduit = address(
+            uint160(
+                uint256(
+                    keccak256(
+                        abi.encodePacked(
+                            bytes1(0xff),
+                            address(_CONDUIT_CONTROLLER),
+                            conduitKey,
+                            _CONDUIT_CREATION_CODE_HASH
+                        )
+                    )
+                )
+            )
+        );
 
-            // Call the conduit and execute bulk transfers.
-            ConduitInterface(conduit).execute(conduitTransfers);
+        // Declare a new array to populate with each token transfer.
+        ConduitTransfer[] memory conduitTransfers = new ConduitTransfer[](
+            totalTransfers
+        );
+
+        // Skip overflow checks: all for loops are indexed starting at zero.
+        unchecked {
+            // Iterate over each transfer.
+            for (uint256 i = 0; i < totalTransfers; ++i) {
+                // Retrieve the transfer in question.
+                TransferHelperItem calldata item = items[i];
+
+                // Create a ConduitTransfer corresponding to each
+                // TransferHelperItem.
+                conduitTransfers[i] = ConduitTransfer(
+                    item.itemType,
+                    item.token,
+                    msg.sender,
+                    recipient,
+                    item.identifier,
+                    item.amount
+                );
+            }
         }
 
-        // Return a magic value indicating that the transfers were performed.
-        magicValue = this.bulkTransfer.selector;
+        // Check if the derived conduit is a contract.
+        if (!_isContract(conduit)) {
+            // If conduit is not a contract, revert with the passed in
+            // conduit key and derived conduit address.
+            revert InvalidConduit(conduitKey, conduit);
+        }
+
+        // Attempt the external call to transfer tokens via the derived conduit.
+        try ConduitInterface(conduit).execute(conduitTransfers) returns (
+            bytes4 conduitMagicValue
+        ) {
+            // Check if the value returned from the external call matches
+            // the conduit `execute` selector.
+            if (
+                conduitMagicValue != ConduitInterface(conduit).execute.selector
+            ) {
+                // If the external call fails, revert with the conduit key
+                // and conduit address.
+                revert InvalidMagicValue(conduitKey, conduit);
+            }
+        } catch (bytes memory data) {
+            // Catch reverts from the external call to the conduit and
+            // "bubble up" the conduit's revert reason if present.
+            if (data.length != 0) {
+                assembly {
+                    returndatacopy(0, 0, returndatasize())
```

Comment:
If we’re being thorough here, we would want to protect against gas exhaustion on loading giant revert reasons into memory; could just throw a generic revert reason if data.length > 256 or something